File: an2_sem4/MPP/Homework/sapt10/java/network/src/main/java/ro/ubbcluj/cs/stefnmubb/festivalsellpoint/network/utils/Stringifier.java
package ro.ubbcluj.cs.stefnmubb.festivalsellpoint.network.utils;

import org.apache.commons.text.StringEscapeUtils;

import java.util.Arrays;

public class Stringifier {

    public static String encode(Object object) {
        if(object instanceof Integer) {
            return Integer.toString((int)object);
        }
        if(object instanceof String){
            return "\""+StringEscapeUtils.escapeJava((String)object)+"\"";
        }

        if(object == null){
            return "null";
        }

        if(object instanceof IStringifiable) {
            return ((IStringifiable) object).encode();
        }

        if(object.getClass().isArray()) {
            if(((Object[])object).length==0)
                return "empty";
            return String.join(",", Arrays.stream(((Object[])object))
                    .map(Stringifier::encode).toList());
        }

        throw new RuntimeException("Cannot stringify object of type "+object.getClass().toString());
    }

    public static<T> T decode(String input, Class<T> tClass){
        return tClass.cast(new SerParser().parse(input));
    }
}

File: an2_sem4/MPP/Homework/sapt9/cs/FestivalSellpoint/FestivalSellpoint.Network/Utils/Stringifier.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace FestivalSellpoint.Network.Utils
{
    public class Stringifier
    {
        public static string Encode(object obj)
        {
            if (obj is int n)
                return n.ToString();

            if (obj is string str)
                return str.ToLiteral();

            if (obj == null)
                return "null";

            if (obj is IStringifiable stringifiable)
            {
                var fields = Reflection.GetAllFields(obj.GetType());
                var propsStr = new List<string>();

                foreach(var f in fields)
                {
                    var value = Encode(f.GetValue(stringifiable));
                    propsStr.Add($"{Reflection.GetSimpleFieldName(f)}={value}");
                }
                return $"{obj.GetType().Name}{{{string.Join(";", propsStr)}}}";                    
            }                

            if(obj.GetType().IsArray)
            {
                var arr = obj as object[];
                if (arr.Length == 0)
                    return "empty";
                return string.Join(",", arr.Select(Encode));
            }

            throw new ArgumentException($"Cannot stringify object of type {obj.GetType()}");
        }


        public static T Decode<T>(string input) => (T)new SerParser().Parse(input);

    }
}
